File: qsort/sort.py
array = [0, 42, 256, 4, 4711, -1337]

def fn(array, offset, last):
    if (offset < last):
        p = fn_part(array, offset, last)
        fn(array, offset, p-1)
        fn(array, p+1, last)

def fn_part(array, offset, last):
    array[last], array[offset] = array[offset], array[last]
    for i in xrange(offset, last+1):
        if array[i] < array[last]:
            array[i], array[offset] = array[offset], array[i] 
            offset += 1
    array[last], array[offset] = array[offset], array[last]
    return offset

fn(array, 0, len(array)-1)
print(array)
    

File: qsort/sort.java
class Sort {
    static int[] array = new int[]{0, 42, 256, 4, 4711, -1337};

    public static void main(String[] args) {
        Sort.fn(0, Sort.array.length-1);

        for(int i : Sort.array) {
            System.out.println(i);
        }
    }

    public static void fn(int offset, int last) {
        if (offset < last) {
            int p = Sort.fn_part(offset, last);
            Sort.fn(offset, p-1);
            Sort.fn(p+1, last);
        }
    }

    public static int fn_part(int offset, int last) {
       int tmp = Sort.array[offset];
       Sort.array[offset] = Sort.array[last];
       Sort.array[last] = tmp;
       for(int i = offset; i <= last; i++) {
            if (Sort.array[i] < Sort.array[last]) {
                tmp = Sort.array[i];
                Sort.array[i] = Sort.array[offset]; 
                Sort.array[offset] = tmp;
                offset++;
            }
       }
       tmp = Sort.array[offset];
       Sort.array[offset] = Sort.array[last];
       Sort.array[last] = tmp;
       return offset;
    }
}
